### ISSUE TITLE
Scope of index variables

### ISSUE THREAD

**Issue by allbits** [jump-dev/JuMP.jl#858]
```julia
m = Model()
@variable(m,x[1:5])
@constraint(m,[m=1:4],x[m]==x[m+1])
```
complains that there's no addconstraint method with an Int64 as the first argument. Should this work?
```julia
@constraint(m,[i=1:4],x[i]==x[i+1])
```
is fine.

**Comment by mlubin** [comment]
> Should this work?

I'm struggling to see the benefit of allowing this to work.

**Comment by allbits** [comment]
> I'm struggling to see the benefit of allowing this to work.

There's probably little benefit, since you always can change the variable. I see this more as a question of clear scoping rules. In 
  `@constraint(m,[m=1:4],x[m]==x[m+1])`
it seems natural to read "[m=1:4]" as a universal quantifier and expect m to be bound to 1:4 in the scope of the restrictions. Or consider:
  `@constraint(m,sum{x[m],m=1:5}==1)`
where it seems natural to expect m to be bound to 1:5 in the scope of the sum.

**Comment by mlubin** [comment]
I'm not opposed to making things more consistent but I don't expect to spend any time fixing this soon. PRs will be considered.

**Comment by mlubin** [comment]
Possibly addressed in Julia 0.7 by https://github.com/JuliaLang/julia/issues/22314

**Comment by mlubin** [comment]
This is not addressed by JuliaLang/julia#22314.
```julia
@constraint(m,[i=1:4],x[i]==x[i+1])
```
is translated essentially to:
```julia
for i in 1:4
    @constraint(m, x[i] == x[i + 1])
end
```
We would need to mangle the iteration variables to avoid the name conflict. This is feasible to do but still not high priority. The only cases I've seen where users run into this issue is when the model is named `m`. We've stopped recommending this name for the model.

**Comment by odow** [comment]
Rather than make this work, we should just throw an error if the index set has the same name as the model.
